### ISSUE TITLE
QNNPACK doesn't respect torch.set_num_threads

### ISSUE THREAD

**Issue by dbalchev** [pytorch/pytorch#45418]
## 🐛 Bug

QNNPACK ops (in particular convolution) uses all CPUs even with `torch.set_num_threads(1)` and `torch.set_num_interop_threads(1)`
## To Reproduce

1. Run the following script
    ```python
    import torch
    
    torch.backends.quantized.engine = 'qnnpack'
    torch.set_num_threads(1)
    torch.set_num_interop_threads(1)
    
    x = torch.quantize_per_tensor(torch.zeros([1, 16, 64, 64]), 0.5, 0, torch.quint8)
    
    conv = torch.nn.intrinsic.quantized.ConvReLU2d(16, 16, 1)
    
    while True:
        conv(x)
    ```
1. Look at the system monitor, `htop` or another util for viewing CPU load

## Expected behavior
The script should use 1 CPU, but it uses all available CPUs. If I use the `fbgemm` backend it behaves as expected.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

```
PyTorch version: 1.6.0+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: Quadro M1200
Nvidia driver version: 440.100
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.6.0+cpu
[conda] numpy                     1.19.2                    <pip>
[conda] torch                     1.6.0+cpu                 <pip>
```

## Additional context
I did some digging about the source of this issue. The issue is that QNNPACK uses `caffe2::pthreadpool_()` as a threadpool (https://github.com/pytorch/pytorch/blob/v1.6.0/aten/src/ATen/native/quantized/cpu/qconv.cpp#L606) and I didn't find a way to set the number of threads in  `caffe2::pthreadpool_()` without recompiling pytorch.

I've posted https://discuss.pytorch.org/t/how-to-set-qnnpack-parallelism/97359, but received no answers.

cc @jerryzh168 @jianyuh @dzhulgakov @raghuramank100 @jamesr66a @vkuzo

**Comment by jerryzh168** [comment]
Asked PoC to respond in discussion forum, closing this for now.

**Comment by ReySadeghi** [comment]
> ## 🐛 Bug
> QNNPACK ops (in particular convolution) uses all CPUs even with `torch.set_num_threads(1)` and `torch.set_num_interop_threads(1)`
> 
> ## To Reproduce
> Steps to reproduce the behavior:
> 
> 1. Run the following script
>    ```python
>    import torch
>    
>    torch.backends.quantized.engine = 'qnnpack'
>    torch.set_num_threads(1)
>    torch.set_num_interop_threads(1)
>    
>    x = torch.quantize_per_tensor(torch.zeros([1, 16, 64, 64]), 0.5, 0, torch.quint8)
>    
>    conv = torch.nn.intrinsic.quantized.ConvReLU2d(16, 16, 1)
>    
>    while True:
>        conv(x)
>    ```
> 2. Look at the system monitor, `htop` or another util for viewing CPU load
> 
> ## Expected behavior
> The script should use 1 CPU, but it uses all available CPUs. If I use the `fbgemm` backend it behaves as expected.
> 
> ## Environment
> Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py) (or fill out the checklist below manually).
> 
> You can get the script and run it with:
> 
> ```
> wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
> # For security purposes, please check the contents of collect_env.py before running it.
> python collect_env.py
> ```
> 
> ```
> PyTorch version: 1.6.0+cpu
> Is debug build: False
> CUDA used to build PyTorch: Could not collect
> ROCM used to build PyTorch: N/A
> 
> OS: Ubuntu 18.04.5 LTS (x86_64)
> GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
> Clang version: Could not collect
> CMake version: version 3.10.2
> 
> Python version: 3.7 (64-bit runtime)
> Is CUDA available: False
> CUDA runtime version: Could not collect
> GPU models and configuration: GPU 0: Quadro M1200
> Nvidia driver version: 440.100
> cuDNN version: Could not collect
> HIP runtime version: N/A
> MIOpen runtime version: N/A
> 
> Versions of relevant libraries:
> [pip3] numpy==1.19.2
> [pip3] torch==1.6.0+cpu
> [conda] numpy                     1.19.2                    <pip>
> [conda] torch                     1.6.0+cpu                 <pip>
> ```
> 
> ## Additional context
> I did some digging about the source of this issue. The issue is that QNNPACK uses `caffe2::pthreadpool_()` as a threadpool (https://github.com/pytorch/pytorch/blob/v1.6.0/aten/src/ATen/native/quantized/cpu/qconv.cpp#L606) and I didn't find a way to set the number of threads in `caffe2::pthreadpool_()` without recompiling pytorch.
> 
> I've posted https://discuss.pytorch.org/t/how-to-set-qnnpack-parallelism/97359, but received no answers.
> 
> cc @jerryzh168 @jianyuh @dzhulgakov @raghuramank100 @jamesr66a @vkuzo

Hi,
I have this problem too. could you find any way to handle it?
please share with me. 
thanks.

**Comment by dbalchev** [comment]
Hey @ReySadeghi 
What version of PyTorch are you using?

**Comment by ReySadeghi** [comment]
> Hey @ReySadeghi What version of PyTorch are you using?

PyTorch: 1.7.1

**Comment by dbalchev** [comment]
Looks like the fix is available from 1.8 on.

Can you update your PyTorch (:

**Comment by ReySadeghi** [comment]
> Looks like the fix is available from 1.8 on.
> 
> Can you update your PyTorch (:

yes, sure

**Comment by ReySadeghi** [comment]
my issue is solved . I could limit number of CPU cores by using docker command:

docker run --cpus 6
